File: elixir/rna-transcription/rna_transcription.exs
defmodule RNATranscription do
  @doc """
  Transcribes a character list representing DNA nucleotides to RNA

  ## Examples

  iex> RNATranscription.to_rna('ACTG')
  'UGAC'
  """
  @spec to_rna([char]) :: [char]
  for {from, to} <- [{'G', 'C'}, {'C', 'G'}, {'T', 'A'}, {'A', 'U'}] do
    def to_rna(unquote(from) ++ tail), do: unquote(to) ++ to_rna(tail)
  end

  def to_rna(''), do: ''
end

File: erlang/rna-transcription/src/rna_transcription.erl
-module(rna_transcription).
-export([to_rna/1]).

to_rna([]) -> [];
to_rna([X | T]) -> [map(X) | to_rna(T)].

map($G) -> $C;
map($C) -> $G;
map($T) -> $A;
map($A) -> $U.
